### DISCUSSION TITLE
Why is the HeaderBag converting header names to lower-case and replacing underscore?

### DISCUSSION THREAD

**Discussion by leberknecht** [symfony/symfony#50658] [Q&A]
From what i got so far, neither underscores nor upper-case letters are invalid in header names (according to RFC822). Why does the `HeaderBag` silently replace those?

Also refs https://stackoverflow.com/questions/22856136/why-do-http-servers-forbid-underscores-in-http-header-names

From my point of view its not good to silently change the header names, i would think that raising an exception if something illegal or (in this case) unwanted happens is better. What about adding something like `$convertHeaderNames = false`, so the user has control over this process.

**Comment by stof** [comment]
Header names are case insensitive in the HTTP spec. That's why HeaderBag converts them to lower case internally (this makes it a lot easier to find an existing header (as we can lowercase the name and then check that key in the array instead of having to check all case permutations as keys in the array).

The conversion of `_` to `-` is because all standardized header names use `-` to separate words while the CGI spec converts those to `_` (with an `HTTP_` prefix as well) to match the naming convention of environment variables.
Note that defining 2 different headers that would only different in using `-` vs `_` would make such HTTP request incompatible with any server using the CGI protocol (which is probably something like 90% of the web, and >99% of servers running PHP).

Note that the HTTP spec even discourages registering a header name containing `_` in the field name registry (i.e. the list of standard field names, i.e. the list of standard headers as headers are fields in that spec). See this quote from section 16.3.1 of the RFC9110 (the current spec defining the semantic of HTTP as RFC822 has been replaced by a set of new specs a few years ago in the httpbis initiative aiming at clarifying all those specs), probably because of this known compatibility issue with the existing infrastructure.

> The requested field name.  It MUST conform to the field-name
      syntax defined in Section 5.1, and it SHOULD be restricted to just
      letters, digits, and hyphen ('-') characters, with the first
      character being a letter.

There is no point adding this `$convertHeaderNames` as disabling the current behavior would violate the HTTP spec (if you talk about the case change) or break compat with CGI (which provides header names using uppercase letter and underscores)

**Comment by leberknecht** [comment]
Thanks for clarification :)
